src/components/GearForm.tsx

```diff
@@ -0,0 +1,187 @@
+
+const useStyles = makeStyles((theme) => ({
+  root: {
+    flexGrow: 1,
+  },
+  menuButton: {
+    marginRight: theme.spacing(2),
+  },
+  title: {
+    flexGrow: 1,
+    color: "white",
+    textAlign: "center",
+    fontSize: "18px",
+  },
+}));
+
+const transition = makeTransition("up");
+
+interface GearFormProps {
+  gear: Equipment[];
+  quantities: {
+    [k: string]: number;
+  };
+  filters: {
+    [k: string]: boolean;
+  };
+  visibleFilters: Set<string>;
+  currentCategory: string;
+  setFieldValue: (field: string, value: number | string | boolean) => void;
+}
+
+const GearForm: FunctionComponent<CalendarUIProps & GearFormProps> = ({
+  dispatch,
+  state,
+  gear,
+  quantities,
+  filters,
+  visibleFilters,
+  currentCategory,
+  setFieldValue,
+}) => {
+  // Constant Declarations
+  const classes = useStyles();
+
+  // State Declarations
+  const [filterDrawerIsOpen, setFilterDrawerIsOpen] = useState(false);
+  const [matchAnyFilter, setMatchAnyFilter] = useState(false);
+  const [searchString, setSearchString] = useState("");
+
+  // Filtering Function to reduce the size of the gear array being passed down
+  function filterItems(
+    gear: Equipment[],
+    filters: { [k: string]: boolean }
+  ): Equipment[] | undefined {
+    let queriedGear: Equipment[] = [];
+    const activeFilters = Object.keys(filters).filter(function (key: string) {
+      return filters[key];
+    });
+    if (searchString !== "") {
+      const queries = searchString.split(",");
+      queriedGear = gear.filter(function (gear) {
+        return queries.some(function (query) {
+          return (
+            gear.description
+              .toLowerCase()
+              .includes(query.toLowerCase().trim()) ||
+            gear.tags.some(function (tag) {
+              return tag.name
+                .toLowerCase()
+                .includes(query.toLowerCase().trim());
+            })
+          );
+        });
+      });
+    } else {
+      queriedGear = gear;
+    }
+    if (matchAnyFilter) {
+      return queriedGear.filter(function (gear) {
+        return activeFilters.some(function (filter) {
+          return gear.tags.some(function (tag) {
+            return tag.name.toLowerCase().includes(filter.toLowerCase().trim());
+          });
+        });
+      });
+    } else {
+      return queriedGear.filter(function (gear) {
+        return activeFilters.every(function (filter) {
+          return gear.tags.some(function (tag) {
+            return tag.name.toLowerCase().includes(filter.toLowerCase().trim());
+          });
+        });
+      });
+    }
+  }
+
+  // Filter Drawer Toggle Function
+  const toggleFilterDrawer = () => (
+    event: React.KeyboardEvent | React.MouseEvent
+  ): void => {
+    if (
+      event.type === "keydown" &&
+      ((event as React.KeyboardEvent).key === "Tab" ||
+        (event as React.KeyboardEvent).key === "Shift")
+    ) {
+      return;
+    }
+    setFilterDrawerIsOpen(!filterDrawerIsOpen);
+  };
```

Comment:
This can be refactored into a separate file.
And let's try to use a reducer rather than multiple loops.

Reply:
There are 3 things going on in `filterItems`:
1. processing the filters: this can be done outside this function; just pass the already processed filters in
1. processing the equipment array with the query string.  NOTE: **don't grab something from the outer context**, make sure you actually pass this string into the function
1. processing the equipment array with the filters

Each of these can be busted out into their own functions, and then we can unit test.  It's hard to unit test the monolith.

Consider
```js
// in /src/utils
const lowerCaseAndTrim = (string) => string.toLowerCase().trim();

// this could be built up from more general functions, but here is a specialized form...
// maybe in /src/calendar/?, maybe in /src/resources/? Equipment?
const lowerCaseAndTrimEquipment = (item) => ({
  ...item,
  description: lowerCaseAndTrim(item.description),
  tags: item.tags.map((tag) => ({ ...tag, name: lowerCaseAndTrim(tag.name) })),
});

// assume that the input to these tests is already lower case and trimmed (see below)
// separating our tests into this isolated function makes it much easier to inspect them
// each item will be subjected to this list of tests
// maybe in /src/calendar/?
const makeTests = (query = "", tags = [], filters = []) => {
  const tests = [];
  if (tags.length)
    tests.push((item) =>
      tags.some((tag) =>
        item.tags.find(({ name }) => name === lowerCaseAndTrim(tag))
      )
    );
  if (filters.length)
    tests.push((item) =>
      filters.some((filter) =>
        item.tags.find(({ name }) => name === lowerCaseAndTrim(filter))
      )
    );
  if (query)
    tests.push((item) =>
      lowerCaseAndTrim(query)
        .split(",")
        .some((token) => item.description.includes(token))
    );
  return tests;
};

// in this "main" function, it's much easier to reason about what we are doing...
// maybe in /src/calendar/?, maybe in /src/resources/? Equipment?
const filterItems = ({ equipment, query, tags, filters, matchAnyFilter }) => {
  const tests = makeTests(query, tags, filters);
  return equipment.reduce((matchingEquipment, item) => {
    const itemUnderTest = lowerCaseAndTrimEquipment(item);
    if (
      (matchAnyFilter && tests.some((test) => test(itemUnderTest))) ||
      (!matchAnyFilter && tests.every((test) => test(itemUnderTest)))
    )
      matchingEquipment.push(item);
    return matchingEquipment;
  }, []);
};

// example usage; this is the only thing you need in GearForm.tsx!!!
const filteredItems = filterItems({equipment, query, tags, filters, matchAnyFilter});
```